agent/src/main/java/org/openremote/agent/protocol/simulator/SimulatorProtocol.java

```diff
@@ -251,56 +286,106 @@ public List<ValueDatapoint<?>> calculatePredictedDatapoints(
+                        return predictedDatapoints;
+                    }
+                    if (schedule.isPresent() && pastUpcoming(schedule.get(), timestamp) && timeSinceOccurrenceStarted >= 0) {
+                        continue;
+                    }
+                    predictedDatapoints.add(new SimulatorReplayDatapoint(timestamp, d.value).toValueDatapoint());
+                }
             }
-            if (!isSingleOccurrence
-                    && (status.equals(PredictedDatapointWindow.BOTH)
-                    || status.equals(PredictedDatapointWindow.NEXT))
-            ) {
-                predictedDatapoints.add(new SimulatorReplayDatapoint((timestamp+occurrenceDuration)*1000, d.value).toValueDatapoint());
+        }
+
+        if (!schedule.map(Schedule::getIsSingleOccurrence).orElse(false)) {
+            if (status.equals(PredictedDatapointWindow.BOTH) || status.equals(PredictedDatapointWindow.NEXT)) {
+                if (schedule.map(s -> s.getUpcoming() == null).orElse(false)) return predictedDatapoints;
+                for (SimulatorReplayDatapoint d : simulatorReplayDatapoints) {
+                    long timeSinceUpcoming = schedule
+                            .map(s -> now - s.getUpcoming().toEpochSecond(ZoneOffset.UTC))
+                            .orElse(timeSinceOccurrenceStarted - 86400L);
+                    OptionalLong delay = Schedule.getDelay(d.timestamp, timeSinceUpcoming, schedule.orElse(null));
+                    if (delay.isEmpty()) {
+                        return predictedDatapoints;
+                    }
+                    long timestamp = (now + delay.getAsLong()) * 1000;
+                    if (schedule.map(s -> s.getHasRecurrenceEnded(timestamp-1)).orElse(false)) {
+                        return predictedDatapoints;
+                    }
+                    predictedDatapoints.add(new SimulatorReplayDatapoint(timestamp, d.value).toValueDatapoint());
+                }
             }
         }
+
         return predictedDatapoints;
     }
 
-    public static class Schedule implements Serializable {
+    /**
+     * Determines for what occurrence window to create predicted datapoints
+     */
+    public enum PredictedDatapointWindow {
+        BOTH,
+        NONE,
+        NEXT
+    }
 
-        @JsonPropertyDescription("Set a start date, if not provided, starts immediately." +
-                " When the replay datapoint timestamp is 0 it will insert it at 00:00.")
-        @JsonSchemaFormat("date-time")
-        @JsonSchemaTypeRemap(type = String.class)
-        protected Date start;
+    public static class Schedule implements Serializable {
 
-        @JsonPropertyDescription("Not implemented, within the recurrence rule you can specify an end date.")
-        @JsonSchemaFormat("date-time")
-        @JsonSchemaTypeRemap(type = String.class)
-        protected Date end;
+        private static class EpochLocalDateTimeDeserializer extends JsonDeserializer<LocalDateTime> {
```

Comment:
Shouldn't all these top-level classes live in their own files?
I notice it was already the case, but it looks quite messy.